File: src/java/org/elastos/hive/vault/backup/credential/CredentialCode.java
package org.elastos.hive.vault.backup.credential;

import org.elastos.hive.DataStorage;
import org.elastos.hive.ServiceEndpoint;
import org.elastos.hive.connection.NodeRPCException;
import org.elastos.hive.connection.auth.CodeFetcher;
import org.elastos.hive.exception.HiveException;
import org.elastos.hive.service.BackupContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class CredentialCode {
	private static final Logger log = LoggerFactory.getLogger(CredentialCode.class);
	private String targetServiceDid;
	private String jwtCode;
	private CodeFetcher remoteResolver;
	private DataStorage storage;

	public CredentialCode(ServiceEndpoint endpoint, BackupContext context) {
		targetServiceDid = context.getParameter("targetServiceDid");
		CodeFetcher remoteResolver = new RemoteResolver(
				endpoint, context, targetServiceDid, context.getParameter("targetAddress"));
		this.remoteResolver = new LocalResolver(endpoint, remoteResolver);
		storage = endpoint.getStorage();
	}

	public String getToken() throws HiveException {
		if (jwtCode != null)
			return jwtCode;

		jwtCode = restoreToken();
		if (jwtCode == null) {
			try {
				jwtCode = remoteResolver.fetch();
			} catch (NodeRPCException e) {
				throw new HiveException(e.getMessage());
			}

			if (jwtCode != null) {
				saveToken(jwtCode);
			}
		}
		return jwtCode;
	}

	private String restoreToken() {
		return storage.loadBackupCredential(targetServiceDid);
	}

	private void saveToken(String jwtCode) {
		storage.storeBackupCredential(targetServiceDid, jwtCode);
	}
}

File: src/restclient/backup/credentialcode.ts
import { DataStorage, Logger, ServiceContext } from "../..";
import { HiveException, NodeRPCException } from "../../exceptions";
import { BackupContext } from "./backupcontext";
import { CodeFetcher } from "./codefetcher";
import { LocalResolver } from "./localresolver";
import { RemoteResolver } from "./remoteresolver";

export class CredentialCode {
    private static LOG = new Logger("CredentialCode");
	private targetServiceDid: string;
	private jwtCode: string;
	private remoteResolver: CodeFetcher;
	private storage: DataStorage;

    constructor(endpoint: ServiceContext,  context: BackupContext) {
        this.targetServiceDid = context.getParameter("targetServiceDid");
        let remoteResolver = new RemoteResolver(
        endpoint, context, this.targetServiceDid, context.getParameter("targetAddress"));
    this.remoteResolver = new LocalResolver(endpoint, remoteResolver);
    this.storage = endpoint.getStorage();
    }
    
    public async getToken(): Promise<string> {
		if (this.jwtCode != null)
			return Promise.resolve(this.jwtCode);

		this.jwtCode = this.restoreToken();
		if (this.jwtCode == null) {
			try {
				this.jwtCode = await this.remoteResolver.fetch();
			} catch (e) {
				throw new HiveException(e.getMessage());
			}

			if (this.jwtCode != null) {
				this.saveToken(this.jwtCode);
			}
		}
		return Promise.resolve(this.jwtCode);
	}

	private  restoreToken(): string {
		return this.storage.loadBackupCredential(this.targetServiceDid);
	}

	private saveToken(jwtCode: string): void {
		this.storage.storeBackupCredential(this.targetServiceDid, jwtCode);
	}
}




